### ISSUE TITLE
Linter always uses /usr/local/bin/flake8

### ISSUE THREAD

**Issue by sjahl** [SublimeLinter/SublimeLinter-flake8#43]
Hi,

At some point in the recent past, sublimelinter started choosing /usr/local/bin/flake8 regardless of the `@python` version I have specified in my project settings. I think the linter used to use the python module (i.e. python -m flake8) with the correct python version.

While I have flake8 installed for both python2 and 3, /usr/local/bin/flake8 only works for the one I've most recently installed it for, since the pip install seems to put the python version in the shebang of that script, like so:

```
#!/usr/local/opt/python/bin/python2.7

# -*- coding: utf-8 -*-
import re
import sys

from flake8.main.cli import main

if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])
    sys.exit(main())
```

This makes flake8 linting not work for python3, or vice versa depending on what I've installed last.

I've tried a few different things, such as setting `@python` inline, changing my shebang lines in scripts to a versioned python (i.e. `#!/usr/local/bin/python3`), but /usr/local/bin/flake8 is always chosen as the linter.

I'm running python 2.7.12 and 3.5.2 installed via Homebrew on OSX 10.11, in case that information is relevant. I'm running the latest build of ST3, and SublimeLinter 3.7.4, and SublimeLinter-flake8 2.0.1

Let me know if I can provide further information.

**Comment by sjahl** [comment]
I was able to resolve this by manually specifying the linter executable with the setting provided in https://github.com/SublimeLinter/SublimeLinter-flake8/pull/38, like so:

```
        "linters":
        {
            "flake8": {
                "executable": "/Users/username/.virtualenvs/myvirtualenv/bin/flake8"
            }
        }
```

Which I can do on a per-project bases (and works well enough for me). I still believe that there's a bug here -- that we can't rely on the executable at /usr/local/bin/flake8 to actually correspond to the version of python we have defined with `@python`.

Perhaps an easy solution is to call flake8 with something like `python<version> -m flake8` as is recommended (for users with multiple python versions installed) in the _Note_ here: https://flake8.readthedocs.io/en/latest/#using-flake8 ?

**Comment by posita** [comment]
> **@sjahl wrote:**
>
> I was able to resolve this by manually specifying the linter executable with the setting provided in #38, like so:
>
```
        "linters":
        {
            "flake8": {
                "executable": "/Users/username/.virtualenvs/myvirtualenv/bin/flake8"
            }
        }
```

This work-around does not work for me. (See SublimeLinter/SublimeLinter3#573.)

> Which I can do on a per-project bases (and works well enough for me).

Just curious, but does [`${project}` expansion](http://www.sublimelinter.com/en/latest/settings.html#setting-tokens) work for you? (Meaning, would it work if you were to place the virtual env inside your project directory and then reference `flake8` by, e.g., `${project}/.virtualenv/bin/flake8`?)

**Comment by ptim** [comment]
@posita the `${project}` expansion does work for me from my project-settings, though my flake8 extensions are not picked up...

```
{
  "SublimeLinter":
  {
    "linters":
    {
      "flake8": {
        "executable": "${project}/.direnv/python-3.6.1/bin/flake8",
      }
    }
  }
}
```

When SublimeLinter debug is true, I see on the console:

```
SublimeLinter: flake8: 0001_initial.py ['/Users/me/source/project/.direnv/python-3.6.1/bin/flake8', '--max-line-length=120', '--jobs=1', '--max-complexity=-1', '-'] 
SublimeLinter: flake8 output:
stdin:44:121: E501 line too long (176 > 120 characters)
```

I want this file (and all migrations) to be ignored via flake8-putty settings (which is working from the command line using the same flake8 binary).

**Comment by brianbruggeman** [comment]
Putting a comment in here so I can come back to the discussion later.

I very much would like to avoid using the `${project}` dir for my virtual environments.  That really goes against the ecosystem (i.e. WORKON folder / multiple projects).  Seems like there's definitely something missing, though.

**Comment by groteworld** [comment]
This type of issue seems to be fixed with SL-flake8 3.0.0. Open if i'm mistaken
